非同期処理(JavaScript)のプログラムを作成

sjabcdefin opened the pull request:
非同期処理(JavaScript)のプログラムを作成いたしました。
該当のソースは、以下となります。

1. コールバック
・エラーなし：sqlite_callback_no_error.js
・エラーあり：sqlite_callback_error.js

2. Promise
・エラーなし：sqlite_promise_no_error.js
・エラーあり：sqlite_promise_error.js

3. async/await
・エラーなし：sqlite_asyncawait_no_error.js
・エラーあり：sqlite_asyncawait_error.js

ご確認お願いいたします。🙇‍♀️

sjabcdefin committed:
install sqlite3

Node.jsからSQLite3を使用するため、sqlite3をインストール

sjabcdefin committed:
Create programs using node-sqlite3 in the form of callback

node-sqlite3をコールバックの形式で使用し、以下の操作についてエラーあり/なしのプログラムをそれぞれ作成
1. テーブルを作成
2. レコードを追加し、自動採番されたIDを標準出力に出力する
3. レコードを取得し、それを標準出力に出力する
4. テーブルを削除する

sjabcdefin committed:
Create programs using Promise wrapper for node-sqlite3 methods(run, all)

node-sqlite3のコールバック形式の関数をラップしてPromise化したプログラムを作成

sjabcdefin committed:
Create programs using async/await

Promise の項目で作った関数を async, await の形式で使用するプログラムを作成

cafedomancer commented on 03.asynchronous/sqlite_callback_noerr.js:
noerr ではなく no err だと思います。それとファイル名に略称を使うのは分かりづらいのでやめましょう。

sjabcdefin replied:
ファイル名について、noerr を no_error、err を error に修正いたしました。

コミット：https://github.com/sjabcdefin/js-practices/pull/3/commits/952081f143f3e4fdfedf904db9f7aaaddf8f9ca4

cafedomancer commented on 03.asynchronous/sqlite_callback_noerr.js:
```diff
@@ -0,0 +1,65 @@
+
+function createTable() {
+  db.run(createTableQuery, insertTitles);
+}
+
+function insertTitles() {
+  console.log("Table created");
+  for (const title of titles) {
+    db.run(insertTitleQuery, [title], displayID);
+  }
+}
+
+function displayID() {
+  console.log(`Record inserted successfully with ID: ${this.lastID}`);
+  insertCount++;
+  if (insertCount === titles.length) {
+    fetchAll();
+  }
+}
+
+function fetchAll() {
+  db.all(selectAllQuery, displayAll);
+}
+
+function displayAll(err, rows) {
+  console.log("All records fetched successfully");
+  for (const row of rows) {
+    console.log(`id:${row.id}, title:${row.title}`);
+  }
+  deleteTable();
+}
+
+function deleteTable() {
+  db.run(deleteTableQuery, closeDB);
+}
+
+function closeDB() {
+  console.log("Table deleted");
+  db.close();
+}
```
いずれのプログラムでも独自の抽象度の関数を定義するのは避けてください。非同期処理の表現方法が移り変わっていく様子をプログラム間で対応づけて理解してほしいのですが、それが実現できなくなってしまうためです。

sjabcdefin replied:
独自の抽象度の関数を削除し、非同期処理をそのまま記述するように修正いたしました。
コミット：https://github.com/sjabcdefin/js-practices/pull/3/commits/77123aa886dc238af1d50e5bd5de2b94198054aa

ひとまず、コールバックベースを修正しております。
そのため、Promiseやasync/awaitでご指摘いただいた以下の項目について、コールバックベースでも修正いたしました。
- https://github.com/sjabcdefin/js-practices/pull/3#discussion_r1929899998
```db.all```について、アロー演算子を使用するように修正いたしました。
コミット：https://github.com/sjabcdefin/js-practices/pull/3/commits/95cd7d9f70c478dfecbe4651c83f37b8101b9ed6

- https://github.com/sjabcdefin/js-practices/pull/3#discussion_r1929900347
パラメータ[]を記述するように修正いたしました。
コミット：https://github.com/sjabcdefin/js-practices/pull/3/commits/9b90b88c076e296500627d63f12aa63f9e0abb25

sjabcdefin replied:
コミット：https://github.com/sjabcdefin/js-practices/pull/3/commits/7f8856a52acc634a34ea2dff5dbb3a8bf627d1ca

Promise、async/awaitについて、非同期処理の表現方法が移り変わっていく様子を分かりやすくするため、main関数を削除いたしました。

cafedomancer commented on 03.asynchronous/sqlite_callback_noerr.js:
```diff
@@ -0,0 +1,65 @@
+
+function insertTitles() {
+  console.log("Table created");
+  for (const title of titles) {
+    db.run(insertTitleQuery, [title], displayID);
+  }
```
コールバックベースの非同期処理を for 文で制御することはできませんよ。

sjabcdefin replied:
すみません。
コールバックベースの非同期処理を制御できる繰り返し処理はない認識であっているでしょうか。💦🙇‍♀️
(複数の非同期処理の待ち合わせはコールバックベースではサポートされていないに該当する。)
レコード追加分、```db.run```を記述いたしました。

コミット：https://github.com/sjabcdefin/js-practices/pull/3/commits/77123aa886dc238af1d50e5bd5de2b94198054aa

cafedomancer replied:
ないわけではないです。再起を用いればコールバックベースでも繰り返し処理を実現することはできます。ただし、node-sqlite3 が提供する関数だけでそれを実現することはできません。

sjabcdefin replied:
ご回答ありがとうございます。😌
かしこまりました。

cafedomancer commented on 03.asynchronous/sqlite_promise_noerr.js:
```diff
@@ -0,0 +1,54 @@
+
+function main() {
+  return runQueryAsync(db, createTableQuery)
+    .then(function () {
```
見づらいので使い分けが必要ない場合は function 式ではなくアロー関数を使ってください。

sjabcdefin replied:
コミット：https://github.com/sjabcdefin/js-practices/pull/3/commits/00979e00294111d4f9206559e81f85b06dac51c7

アロー関数を使用するように修正いたしました。
なお、sqlite_utils.js の ```runQueryAsync```関数の```db.run```のみコールバック関数内で```this```を使用しているため、```function () { ... }```式を使用しています。🙏🙇‍♀️
https://github.com/sjabcdefin/js-practices/pull/3#discussion_r1947719398 でのやり取り結果で、修正が必要な場合は、修正いたします。🙇‍♀️

cafedomancer commented on 03.asynchronous/sqlite_promise_noerr.js:
```diff
@@ -0,0 +1,54 @@
+  return runQueryAsync(db, createTableQuery)
+    .then(function () {
+      console.log("Table created");
+      const requests = titles.map(function (title) {
+        return runQueryAsync(db, insertTitleQuery, [title]).then(
+          function (result) {
+            console.log(
+              `Record inserted successfully with ID: ${result.lastID}`,
+            );
+          },
+        );
+      });
+      return Promise.all(requests);
```
各処理は並列ではなく直列に実行してください。

sjabcdefin replied:
コミット：https://github.com/sjabcdefin/js-practices/pull/3/commits/68c7e4298dc4cd675553266fbe0994044f769edb

並列ではなく、直列で実施するように修正いたしました。

cafedomancer commented on 03.asynchronous/sqlite_promise_noerr.js:
```diff
@@ -0,0 +1,54 @@
+      console.log("Table deleted");
+    })
+    .finally(function () {
+      db.close();
```
close もコールバックベースの非同期処理を行う関数なので Promise として扱えるようにしてください。

sjabcdefin replied:
コミット：https://github.com/sjabcdefin/js-practices/pull/3/commits/233099ebd3cab4dfca4eeeaa1e3cdc70132195a6

close について Promise として扱えるように修正いたしました。

cafedomancer commented on 03.asynchronous/sqlite_utils.js:
```diff
@@ -0,0 +1,23 @@
+export function runQueryAsync(db, query, params = []) {
```
質問ですが、`params` にデフォルト引数を設定しているのは何のためですか？

cafedomancer replied:
質問に対する回答がないです。

sjabcdefin replied:
失礼いたしました。🙏🙇‍♀️
node-sqlite3 の API ドキュメント https://github.com/TryGhost/node-sqlite3/wiki/API
に、
以下の記載があったため、第2引数に```[]```を指定したほうがよいと思い、デフォルト引数として指定いたしました。
認識に誤りがありましたら、教えていただけると幸いです。


>run(sql [, param, ...] [, callback])
...
In case you want to keep the callback as the 3rd parameter, you should set param to "[]" ( Empty Array ) as per https://github.com/mapbox/node-sqlite3/issues/116

sjabcdefin replied:
コミット：https://github.com/sjabcdefin/js-practices/pull/3/commits/56986a0d819d367cadb944dd3230ba34dc590a52

以下の理由のため、[]を削除(省略)いたしました。
・#116 の問題は発生していない
・クエリに```?```がある場合は、paramsが必ず指定され、```?```がない場合、paramsは```undefined```になるが、クエリの実行には問題ない。

cafedomancer commented on 03.asynchronous/sqlite_utils.js:
```diff
@@ -0,0 +1,23 @@
+  });
+}
+
+export function allQueryAsync(db, query) {
```
元の all 関数もパラメータを受け付けるはずです。

sjabcdefin replied:
コミット：https://github.com/sjabcdefin/js-practices/pull/3/commits/f23f1de1e25363539e50e973f74e785e8ea7e3f7

```allQueryAsync```関数及び```db.all```の引数に、```params```を追加いたしました。

cafedomancer commented on 03.asynchronous/sqlite_callback_noerr.js:
```diff
@@ -0,0 +1,65 @@
+    title TEXT NOT NULL UNIQUE
+  )
+`;
+const insertTitleQuery = `INSERT INTO books (title) VALUES (?)`;
+const selectAllQuery = `SELECT * FROM books`;
+const deleteTableQuery = `DROP TABLE books`;
```
文字列中に変数を埋め込んでいるわけではないのにテンプレート文字列が使われています。

sjabcdefin replied:
すみません。以前のプラクティスでも同様の指摘をいただいておりました。😭🙏
ダブルクウォーテーションを使用するように修正いたしました。

コミット：https://github.com/sjabcdefin/js-practices/pull/3/commits/20471286e0736da50888936111d99064ad858971

cafedomancer commented on 03.asynchronous/sqlite_callback_noerr.js:
```diff
@@ -0,0 +1,65 @@
+sqlite3.verbose();
+const db = new sqlite3.Database(":memory:");
+
+const createTableQuery = `
+  CREATE TABLE books (
+    id INTEGER PRIMARY KEY AUTOINCREMENT,
+    title TEXT NOT NULL UNIQUE
+  )
+`;
```
これだと改行分やインデント分も文字列に含まれるので、以下のような文字列になってしまいます。

```
<改行>
<空白><空白>CREATE TABLE books (
<空白><空白><空白><空白>id INTEGER PRIMARY KEY AUTOINCREMENT,
<空白><空白><空白><空白>title TEXT NOT NULL UNIQUE
<空白><空白>)
```

sjabcdefin replied:
特別長いSQL文ではないため、
可読性を上げるための余分な改行やインデントは不要と考え、一行で書くようにいたしました。
コミット：https://github.com/sjabcdefin/js-practices/pull/3/commits/f1c8683db2d169de8bc27019b9790b6c3aa3f2b6

cafedomancer commented on 03.asynchronous/sqlite_asyncawait_err.js:
```diff
@@ -0,0 +1,49 @@
+    try {
+      const result = await runQueryAsync(db, insertTitleQuery, [title]);
+      console.log(`Record inserted successfully with ID: ${result.lastID}`);
+    } catch (err) {
+      console.error(`Error occurred while inserting record: ${err.message}`);
+    }
```
catch では全ての例外が無条件に捕捉されるので、この実装だと例外が握り潰されています。例外は握り潰さずに、捕捉したい例外のみを捕捉するようにしてください。

sjabcdefin replied:
コミット：https://github.com/sjabcdefin/js-practices/pull/3/commits/12973cf0e7d475400c6fdc40cb7d56f1b2771dc3

```err.code```の判定処理を追加し、捕捉したい例外のみを捕捉するように修正いたしました。
こちらのご指摘は、Promiseにも該当すると思い、Promiseも修正しております。🙏

cafedomancer replied:
例外が引き続き握り潰されています。

sjabcdefin replied:
コミット：https://github.com/sjabcdefin/js-practices/pull/3/commits/eddc819652bc0461b5408f406132f78f7dc09730

想定しているエラー（SQLITE_CONSTRAINT、SQLITE_ERROR）以外についても握りつぶさず、捕捉するように修正いたしました。

sjabcdefin committed:
Rename files to use "error" instead of "err" and "no_error" instead of "noerr"

可読性を向上させるために、ファイル名を変更

sjabcdefin committed:
Use double quotes instead of backquotes

文字列中に変数を埋め込んでいるわけではないため、バッククウォートではなくダブルクウォーテーションを使用

sjabcdefin committed:
Refactor SQL queries by removing unnecessary line breaks and indentation

SQLクエリから不要な改行とインデントを削除し、一行に統一

sjabcdefin committed:
Removed abstraction in asynchronous processing and rewrote it in a direct form

独自の抽象関数を使用せず、非同期処理を直接記述する形式に修正
- コールバックベースの処理をそのまま記述
- 非同期処理の流れが分かりやすくなるよう変更
- for文を使用せず、レコード追加を追加分そのまま記述

sjabcdefin committed:
Writes an empty array [] even if no parameters are specified

パラメータの指定がない場合も[]を記載するように修正

sjabcdefin committed:
Fixed db.all to use arrow function instead of function expression

見やすくなるように、アロー演算子を使用するように修正

cafedomancer commented on 03.asynchronous/sqlite_callback_no_error.js:
```diff
@@ -0,0 +1,37 @@
+
+const createTableQuery =
+  "CREATE TABLE books (id INTEGER PRIMARY KEY AUTOINCREMENT, title TEXT NOT NULL UNIQUE);";
+const insertTitleQuery = "INSERT INTO books (title) VALUES (?)";
```
insert するのは title カラムではなく books テーブルのレコードではないですかね。

sjabcdefin replied:
```insertTitleQuery```を```insertBookRecordQuery```に変更いたしました。
コミット：https://github.com/sjabcdefin/js-practices/pull/3/commits/b7d6d4db55da0d51dbec9a106ac3e52a45a0889f

cafedomancer commented on 03.asynchronous/sqlite_callback_no_error.js:
```diff
@@ -0,0 +1,37 @@
+const titles = ["I Am a Cat", "KOKORO", "SANSHIRO"];
+
+db.run(createTableQuery, [], function () {
+  console.log("Table created");
```
create は他動詞なので文言が文法的におかしそうです。

sjabcdefin replied:
Table was created successfully に修正いたしました。
create は他動詞であるため、受動態とし、成功したことが分かりやすいように、successfully を追加いたしました。

また、insert, fetch, delete も他動詞であるため、同様の修正をいたしました。

コミット：https://github.com/sjabcdefin/js-practices/pull/3/commits/526b0a37a239919c3344f14ea858a498fbd1b5dc

(追記)
Table was deleted successfully について、successfully にスペルミスがあることに気付いたため、修正いたしました。
コミット：https://github.com/sjabcdefin/js-practices/pull/3/commits/45f478fcff0b9e509c67cf64125424095cb2507a

cafedomancer commented on 03.asynchronous/sqlite_callback_no_error.js:
```diff
@@ -0,0 +1,37 @@
+
+const titles = ["I Am a Cat", "KOKORO", "SANSHIRO"];
+
+db.run(createTableQuery, [], function () {
```
修正されていないです。https://github.com/sjabcdefin/js-practices/pull/3#discussion_r1929899998 それから、パラメータの値として `[]` を渡しているのはなぜでしょうか？

sjabcdefin replied:
node-sqlite3 の API ドキュメント https://github.com/TryGhost/node-sqlite3/wiki/API

失礼いたしました。
node-sqlite3 の API ドキュメントに、以下の記載があったため、
```db.run```では、```function () { ... }```を使用したほうがよいのかなと思ったのですが、
そういうわけではないでしょうか。🙏🙇‍♀️
>run(sql [, param, ...] [, callback])
...
you must use an old-school function () { ... } style callback rather than a lambda function, otherwise this.lastID and this.changes will be undefined.

パラメータの値として第2引数に```[]```を渡している理由ですが、
node-sqlite3 の API ドキュメントに、以下の記載があったため、第2引数に```[]```を指定いたしました。
認識誤っていましたら、教えていただけると幸いです。
>run(sql [, param, ...] [, callback])
...
In case you want to keep the callback as the 3rd parameter, you should set param to "[]" ( Empty Array ) as per [issue #116](https://github.com/mapbox/node-sqlite3/issues/116)

cafedomancer replied:
> 失礼いたしました。
> node-sqlite3 の API ドキュメントに、以下の記載があったため、
> `db.run`では、`function () { ... }`を使用したほうがよいのかなと思ったのですが、
> そういうわけではないでしょうか。🙏🙇‍♀️

確認ですが、コールバック関数内で `this` を参照しているのですか？

> パラメータの値として第2引数に[]を渡している理由ですが、
node-sqlite3 の API ドキュメントに、以下の記載があったため、第2引数に[]を指定いたしました。
認識誤っていましたら、教えていただけると幸いです。

確認ですが、#116 の issue の問題が発生しているのですか？

sjabcdefin replied:
> 確認ですが、コールバック関数内で this を参照しているのですか？

コールバック関数内で```this```を参照しています。🙏🙇‍♀️
```this```を使用しないほうが好ましいのでしょうか。

> 確認ですが、#116 の issue の問題が発生しているのですか？

コミット：https://github.com/sjabcdefin/js-practices/pull/3/commits/26e8ba8369815793a9052135a15acde58e2894be

#116 の問題は発生していません。
そのため、パラメータの指定が不要の場合は、```[]```を削除(省略)いたしました。

cafedomancer replied:
> コールバック関数内で`this`を参照しています。🙏🙇‍♀️ `this`を使用しないほうが好ましいのでしょうか。

`this` がどの関数に対応するものなのかをよく見直してください。

sjabcdefin replied:
コミット：https://github.com/sjabcdefin/js-practices/pull/3/commits/6ef5baf09eb9ce69a14d3170f988ec35aec1fc96

```this```を使う```db.run```のコールバックはfunction形式のままとし、
```this```を使わない```db.run```のコールバックはアロー関数に修正いたしました。

すみません。ご指摘の内容の修正について、上記で合っているでしょうか。💦
合っていない場合、再度ご指摘いただきますようお願いいたします。🙏🙇‍♀️

sjabcdefin committed:
Rename insertTitleQuery to insertBookRecordQuery

titleカラムを挿入するのではなく、booksテーブルに新しいレコードを追加することを明確にするため。

sjabcdefin committed:
Change log messages to passive voice

正しい文法とするため、他動詞(create, insert, fetch, delete)について、受動態に変更。

sjabcdefin commented:
> https://github.com/sjabcdefin/js-practices/pull/3#discussion_r1934085790
質問に対する回答がないです。

失礼いたしました。
https://github.com/sjabcdefin/js-practices/pull/3#discussion_r1938235543 に返信しております。
ご確認お願いいたします。

sjabcdefin committed:
Correct spelling mistakes

successfullyのスペルミスを修正

cafedomancer commented on 03.asynchronous/sqlite_callback_no_error.js:
```diff
@@ -0,0 +1,37 @@
+
+const createTableQuery =
+  "CREATE TABLE books (id INTEGER PRIMARY KEY AUTOINCREMENT, title TEXT NOT NULL UNIQUE);";
+const insertBookRecordQuery = "INSERT INTO books (title) VALUES (?)";
+const selectAllQuery = "SELECT * FROM books";
```
この 2 つの変数名に一貫性がないです。

sjabcdefin replied:
コミット：https://github.com/sjabcdefin/js-practices/pull/3/commits/edf50f3e1c972a2a2fe129555683997ac4d9e47d

```insertBookRecordQuery```を```insertRecordQuery```に修正いたしました。

cafedomancer replied:
問題が解決していないように思いますが。

sjabcdefin replied:
コミット：https://github.com/sjabcdefin/js-practices/pull/3/commits/d722ec9f537125f4021528de2ae70a891ed475b7

「行を挿入するクエリ」として、```insertRecordQuery```を```insertRowQuery```に、
「全ての行を選択するクエリ」として、```selectAllQuery```を```selectAllRowsQuery```に修正いたしました。
行挿入の方が適切かと考え、RecordをRowに変更しております。

cafedomancer commented on 03.asynchronous/sqlite_callback_no_error.js:
```diff
@@ -0,0 +1,37 @@
+  "CREATE TABLE books (id INTEGER PRIMARY KEY AUTOINCREMENT, title TEXT NOT NULL UNIQUE);";
+const insertBookRecordQuery = "INSERT INTO books (title) VALUES (?)";
+const selectAllQuery = "SELECT * FROM books";
+const deleteTableQuery = "DROP TABLE books";
```
SQL と対応づけるなら drop にしたほうがよいと思います。他の変数名は SQL の命令と対応づいているのにこの変数だけ対応づいていない点も気になります。

sjabcdefin replied:
コミット：https://github.com/sjabcdefin/js-practices/pull/3/commits/ef60c8e40a6ba8224f7fc194674c0c89a66b1207

```deleteTableQuery```を```dropTableQuery```に修正いたしました。

sjabcdefin committed:
Rename variable name in record insertquery

他の変数名との一貫性を取るため、レコード挿入クエリの変数名を変更

sjabcdefin committed:
Rename variable name in table dletion query

SQL と対応づけるため、テーブル削除クエリの変数名を変更

sjabcdefin committed:
Remove empty array when parameters are not spericied

パラメータが指定されていない場合、空配列を省略

sjabcdefin committed:
Remove main function

非同期処理の表現方法が移り変わっていく様子を分かりやすくするため

sjabcdefin committed:
Fixed to use arrow function instead of function expression

可読性向上のため

sjabcdefin committed:
Fixed to insert records serially instead of parallel

レコード挿入について、並列ではなく直列で実施されるように修正

sjabcdefin committed:
Refactor db.close() to be an asynchronous function and move it to sqlite_utils.js as closeQueryAsync

closeも非同期処理を行う関数であるため

sjabcdefin committed:
Remove the default value [] for the params argument in runQueryAsync

クエリに?が含まれていない場合、パラメータを指定する必要はないため

sjabcdefin committed:
Add params argument to db.all

元の all 関数もパラメータを受け付けるため

sjabcdefin committed:
Fixed to catch only the specified exception

捕捉したい例外のみを捕捉するように修正

cafedomancer commented on 03.asynchronous/sqlite_callback_no_error.js:
```diff
@@ -0,0 +1,37 @@
+          for (const row of rows) {
+            console.log(`id:${row.id}, title:${row.title}`);
+          }
+          db.run(dropTableQuery, function () {
```
修正されていないです。https://github.com/sjabcdefin/js-practices/pull/3#discussion_r1929899998

sjabcdefin replied:
コミット：https://github.com/sjabcdefin/js-practices/pull/3/commits/6ef5baf09eb9ce69a14d3170f988ec35aec1fc96

```this```を使わない```db.run```のコールバックはアロー関数に修正いたしました。

cafedomancer commented on 03.asynchronous/sqlite_utils.js:
```diff
@@ -0,0 +1,35 @@
+  });
+}
+
+export function closeQueryAsync(db) {
```
close は query ではないし、query を close するわけでもないです。

sjabcdefin replied:
コミット：https://github.com/sjabcdefin/js-practices/pull/3/commits/9601d227d25915dca9955cf25a985a2f09eee929

「データベース接続を閉じる」処理として、```closeDatabaseAsync```に修正いたしました。

cafedomancer commented on 03.asynchronous/sqlite_promise_no_error.js:
```diff
@@ -0,0 +1,50 @@
+  .then(() => {
+    console.log("Table was deleted successfully");
+  })
+  .finally(() => {
+    return closeQueryAsync(db);
+  });
```
アロー関数の中身が return しかないときは簡潔文体を使ってください。

sjabcdefin replied:
コミット：https://github.com/sjabcdefin/js-practices/pull/3/commits/a57a47a931f170d6784e63cd68ec22b13b9eed32

アロー関数の中身がreturn しかないときは、簡潔文体を使うよう修正いたしました。
sqlite_promise_error.jsも同様に修正しております。

cafedomancer commented on 03.asynchronous/sqlite_asyncawait_no_error.js:
```diff
@@ -0,0 +1,40 @@
+
+const titles = ["I Am a Cat", "KOKORO", "SANSHIRO"];
+
+(async () => {
+  await runQueryAsync(db, createTableQuery);
+  console.log("Table was created successfully");
+
+  for (const title of titles) {
+    const result = await runQueryAsync(db, insertRecordQuery, [title]);
+    console.log(`Record was inserted successfully with ID: ${result.lastID}`);
+  }
+
+  const rows = await allQueryAsync(db, selectAllQuery);
+  console.log("All records were fetched successfully");
+  for (const row of rows) {
+    console.log(`id:${row.id}, title:${row.title}`);
+  }
+
+  await runQueryAsync(db, dropTableQuery);
+  console.log("Table was deleted successfully");
+
+  await closeQueryAsync(db);
+})();
```
即時実行関数式ではなく Top-level await を使ってください。

sjabcdefin replied:
コミット：https://github.com/sjabcdefin/js-practices/pull/3/commits/8ec4514709055948f888d677766ea0257544a465

Top-level await に修正いたしました。
sqlite_promise_error.jsも同様に修正しております。

cafedomancer commented on 03.asynchronous/sqlite_asyncawait_no_error.js:
```diff
@@ -0,0 +1,40 @@
+  await runQueryAsync(db, dropTableQuery);
+  console.log("Table was deleted successfully");
+
+  await closeQueryAsync(db);
```
Promise のプログラムではこの処理は finally で実行されていたはずです。

sjabcdefin replied:
コミット：https://github.com/sjabcdefin/js-practices/pull/3/commits/92ebeb136d7b3aa7bbafc0e0984d0653e1e6acdb

```closeDatabaseAsync```をfinallyブロックで実行するように修正いたしました。
sqlite_promise_error.jsも同様に修正しております。

cafedomancer commented on 03.asynchronous/sqlite_asyncawait_error.js:
```diff
@@ -0,0 +1,52 @@
+      const result = await runQueryAsync(db, insertRecordQuery, [title]);
+      console.log(`Record was inserted successfully with ID: ${result.lastID}`);
+    } catch (err) {
+      if (err.code === "SQLITE_CONSTRAINT") {
```
この条件だと `err` 変数にプロパティアクセスできない値が含まれていると `err.code` の部分でエラーが起きます。それから Error オブジェクトでなくても条件を満たしていればこの条件にマッチしてしまいます。例えば以下のようなオブジェクトが例外として送出される場合です。

```js
{ code: "SQLITE_CONSTRAINT" }
```

sjabcdefin replied:
コミット：https://github.com/sjabcdefin/js-practices/pull/3/commits/eddc819652bc0461b5408f406132f78f7dc09730

判定に```err instanceof Error```を追加し、
err変数がプロパティアクセスできない値やErrorオブジェクトではない場合は、想定外のエラーとなるように修正いたしました。

sjabcdefin committed:
Fixed to use arrow functions for db.run where this is not used

db.run のコールバック関数内で this を使用しない箇所について、可読性向上のために function 式からアロー関数へ変更

sjabcdefin committed:
Rename variable name in insertquery and selectquery

レコード挿入クエリとレコード選択クエリの変数名について、一貫性を取るため

sjabcdefin committed:
Rename closeQueryAsync to closeDatabaseAsync

closeQueryAsync という関数名を、実際に行っている「データベース接続を閉じる」処理を反映するために closeDatabaseAsync に変更。

sjabcdefin committed:
Fixed to use concise writing when arrow functions only contain return

可読性向上のため、アロー関数にreturnのみが含まれる場合は簡潔文体を使用

sjabcdefin committed:
Refactor to be Top-level await

即時実行関数式ではなく Top-level await に修正

sjabcdefin committed:
Move DB close processing to finally block

DBクローズ処理をfinallyブロックに移動

sjabcdefin committed:
Handle specific SQLite errors and rethrow unexpected ones

想定したエラー(SQLITE_CONSTRAINT, SQLITE_ERROR)のみを適切に処理し、意図しない例外の握りつぶしを防ぐため。

cafedomancer commented on 03.asynchronous/sqlite_asyncawait_no_error.js:
コメントし忘れていましたが、asyncawait ではなく async/await なので 2 語です。

sjabcdefin replied:
コミット：https://github.com/sjabcdefin/js-practices/pull/3/commits/4d22ca968b1d89cdb00dd1b70a167de43d559594

asyncawait のファイル名を async_await に修正いたしました。
sqlite_async_await_error.js についても同様に修正しております。

cafedomancer commented on 03.asynchronous/sqlite_asyncawait_error.js:
```diff
@@ -0,0 +1,58 @@
+      if (err instanceof Error && err.code === "SQLITE_CONSTRAINT") {
+        console.error(`Error occurred while inserting record: ${err.message}`);
+      } else {
+        throw new Error("Unexpected error occured while inserting record");
```
元の例外と同じものを再送出してください。でないと元の例外が伝播しなくなってしまいます。

sjabcdefin replied:
コミット：https://github.com/sjabcdefin/js-practices/pull/3/commits/1726f9d640641fd1c5b115e7171733bc8e136b9e

元の例外を再送出するように修正いたしました。

cafedomancer commented on 03.asynchronous/sqlite_asyncawait_error.js:
```diff
@@ -0,0 +1,58 @@
+
+  await runQueryAsync(db, dropTableQuery);
+  console.log("Table was deleted successfully");
+} catch (err) {
+  console.error(err);
```
今回のプログラムでは、エラーが起こる想定の部分に対してのみエラーハンドリングを行ってください。この catch に対応するものはないはずです。

sjabcdefin replied:
コミット：https://github.com/sjabcdefin/js-practices/pull/3/commits/64325e23d7c00197d93c2d24dde75381321fe951

エラーが起こる想定の部分に対してのみエラーハンドリングをするため、この catch ブロック は削除いたしました。

sjabcdefin committed:
Rename files  from asyncawait to async_await

asyncawait ではなく async/await の 2 語であるため

sjabcdefin committed:
Fixed to rethrow the original exception

元の例外を伝播させるため

sjabcdefin committed:
Fixed to handle only expected errors

今回のプログラムでは、エラーが起こる想定の部分に対してのみエラーハンドリングを行うため
